File: legacy/python/disk_agent/report.py
from __future__ import annotations

from pathlib import Path
from typing import Iterable, List

from .models import DirectoryUsage, Snapshot, load_snapshot
from .snapshot import SNAPSHOT_DIR


def format_bytes(value: int | None, signed: bool = False) -> str:
    if value is None:
        return "unavailable"
    sign = ""
    if signed:
        sign = "+" if value >= 0 else "-"
        value = abs(value)
    units = ["B", "K", "M", "G", "T", "P"]
    amount = float(value)
    for unit in units:
        if amount < 1024 or unit == units[-1]:
            rendered = f"{amount:.1f}".rstrip("0").rstrip(".")
            return f"{sign}{rendered}{unit}"
        amount /= 1024
    return f"{sign}{value}B"


def snapshot_paths(directory: Path = SNAPSHOT_DIR) -> List[Path]:
    return sorted(directory.glob("????-??-??.json")) if directory.exists() else []


def latest_snapshot(directory: Path = SNAPSHOT_DIR) -> Snapshot:
    paths = snapshot_paths(directory)
    if not paths:
        raise RuntimeError("no snapshots found; run 'disk-agent snapshot' first")
    return load_snapshot(paths[-1])


def _top_consumers(snapshot: Snapshot, limit: int = 5) -> Iterable[DirectoryUsage]:
    # Immediate children of home provide a stable, non-overlapping summary.
    return sorted(
        (item for item in snapshot.home_usage if item.path != "~" and item.path.count("/") == 1),
        key=lambda item: item.bytes,
        reverse=True,
    )[:limit]


def render_report(snapshot: Snapshot) -> str:
    fs = snapshot.filesystem
    lines = [
        f"Filesystem usage: {fs.used_percent}% ({format_bytes(fs.used_bytes)} of {format_bytes(fs.total_bytes)})",
        "",
        "Top consumers:",
        "",
    ]
    consumers = list(_top_consumers(snapshot))
    lines.extend(f"{format_bytes(item.bytes)} {item.path}" for item in consumers)
    if not consumers:
        lines.append("No directory data available.")

    lines.extend(["", "Podman:", ""])
    if snapshot.podman.available:
        lines.extend(
            [
                f"Images: {format_bytes(snapshot.podman.images_bytes)}",
                f"Containers: {format_bytes(snapshot.podman.containers_bytes)}",
                f"Volumes: {format_bytes(snapshot.podman.volumes_bytes)}",
            ]
        )
    else:
        lines.append(f"Unavailable ({snapshot.podman.error or 'unknown error'}).")

    lines.extend(["", "Largest directories:", ""])
    lines.extend(f"{format_bytes(item.bytes)} {item.path}" for item in snapshot.largest_directories[:10])
    lines.extend(["", "Assessment:"])
    if fs.used_percent >= 90:
        lines.append("Disk usage is critical.")
    elif fs.used_percent >= 80:
        lines.append("Disk usage is elevated.")
    else:
        lines.append("No action required.")
    return "\n".join(lines)


def report_command() -> str:
    return render_report(latest_snapshot())

File: src/report.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{anyhow, Result};

use crate::json::load_snapshot;
use crate::models::{DirectoryUsage, Snapshot};
use crate::output::format_bytes;
use crate::paths;

pub fn snapshot_paths(directory: &Path) -> Result<Vec<PathBuf>> {
    if !directory.exists() {
        return Ok(Vec::new());
    }

    let mut paths = Vec::new();
    for entry in fs::read_dir(directory)? {
        let path = entry?.path();
        if is_snapshot_name(&path) {
            paths.push(path);
        }
    }
    paths.sort();
    Ok(paths)
}

pub fn latest_snapshot_from(directory: &Path) -> Result<Snapshot> {
    Ok(latest_snapshot_with_path_from(directory)?.snapshot)
}

pub fn latest_snapshot_with_path_from(directory: &Path) -> Result<LoadedSnapshot> {
    let paths = snapshot_paths(directory)?;
    let Some(path) = paths.last() else {
        return Err(anyhow!(
            "no snapshots found; run 'disk-agent snapshot' first"
        ));
    };
    Ok(LoadedSnapshot {
        snapshot: load_snapshot(path)?,
        path: path.clone(),
    })
}

pub struct LoadedSnapshot {
    pub snapshot: Snapshot,
    pub path: PathBuf,
}

pub fn render_report(snapshot: &Snapshot) -> String {
    let fs = &snapshot.filesystem;
    let mut lines = vec![
        format!(
            "Filesystem usage: {}% ({} of {})",
            fs.used_percent,
            format_bytes(Some(fs.used_bytes), false),
            format_bytes(Some(fs.total_bytes), false)
        ),
        String::new(),
        "Top consumers:".to_string(),
        String::new(),
    ];

    let consumers = top_consumers(snapshot, 5);
    if consumers.is_empty() {
        lines.push("No directory data available.".to_string());
    } else {
        lines.extend(
            consumers
                .iter()
                .map(|item| format!("{} {}", format_bytes(Some(item.bytes), false), item.path)),
        );
    }

    lines.extend([String::new(), "Podman:".to_string(), String::new()]);
    if snapshot.podman.available {
        lines.extend([
            format!(
                "Images: {}",
                format_bytes(snapshot.podman.images_bytes, false)
            ),
            format!(
                "Containers: {}",
                format_bytes(snapshot.podman.containers_bytes, false)
            ),
            format!(
                "Volumes: {}",
                format_bytes(snapshot.podman.volumes_bytes, false)
            ),
        ]);
    } else {
        lines.push(format!(
            "Unavailable ({}).",
            snapshot.podman.error.as_deref().unwrap_or("unknown error")
        ));
    }

    lines.extend([
        String::new(),
        "Largest directories:".to_string(),
        String::new(),
    ]);
    lines.extend(
        snapshot
            .largest_directories
            .iter()
            .take(10)
            .map(|item| format!("{} {}", format_bytes(Some(item.bytes), false), item.path)),
    );
    lines.extend([String::new(), "Assessment:".to_string()]);
    if fs.used_percent >= 90 {
        lines.push("Disk usage is critical.".to_string());
    } else if fs.used_percent >= 80 {
        lines.push("Disk usage is elevated.".to_string());
    } else {
        lines.push("No action required.".to_string());
    }

    lines.join("\n")
}

pub fn report_command(refresh: bool) -> Result<String> {
    let directory = paths::snapshot_dir()?;
    if refresh {
        let snapshot = crate::snapshot::collect_snapshot()?;
        let path = crate::snapshot::save_snapshot(&snapshot, &directory)?;
        return Ok(render_report_with_metadata(&snapshot, &path));
    }

    let loaded = latest_snapshot_with_path_from(&directory)?;
    Ok(render_report_with_metadata(&loaded.snapshot, &loaded.path))
}

pub fn render_report_with_metadata(snapshot: &Snapshot, path: &Path) -> String {
    format!(
        "Snapshot: saved {}\nSource: {}\n\n{}",
        snapshot.timestamp,
        path.display(),
        render_report(snapshot)
    )
}

fn top_consumers(snapshot: &Snapshot, limit: usize) -> Vec<&DirectoryUsage> {
    let mut values = snapshot
        .home_usage
        .iter()
        .filter(|item| item.path != "~" && item.path.matches('/').count() == 1)
        .collect::<Vec<_>>();
    values.sort_by(|left, right| right.bytes.cmp(&left.bytes));
    values.truncate(limit);
    values
}

fn is_snapshot_name(path: &Path) -> bool {
    let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
        return false;
    };
    let bytes = name.as_bytes();
    bytes.len() == 15
        && bytes[4] == b'-'
        && bytes[7] == b'-'
        && &bytes[10..] == b".json"
        && bytes[..4].iter().all(u8::is_ascii_digit)
        && bytes[5..7].iter().all(u8::is_ascii_digit)
        && bytes[8..10].iter().all(u8::is_ascii_digit)
}
